20_2 datagrid: improve the rendering optimization article (#1729)
DevExpress/devextreme-documentation: concepts/05 UI Components/DataGrid/10 Enhance Performance on Large Datasets/030 Rendering Optimization.md

> If you work with a large data set, follow the recommendations below to optimize rendering performance:

Reviewer: ```suggestion
If you work with a large data set, follow the recommendations below to improve rendering performance:
```

> If you work with a large data set, apply the following recommendations to optimize rendering performance:

Reviewer 2: The phrase "apply recommendation" sounds strange to me. I don't think these words are commonly used together. I'd say "follow recommendations", maybe "consider recommendations". 

```suggestion
If you work with a large data set, follow the recommendations below to optimize rendering performance:
```

> * **Render rows and columns of the current viewport only**             

Reviewer 3: ```suggestion
* **Render only visible rows and columns**             
```

> The default behavior is to render all loaded rows and columns at once. To render those of the current viewport only, set the [rowRenderingMode](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode) and [columnRenderingMode](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode) properties to *"virtual"*. You should do this for rows if [paging was disabled](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled) or the **paging**.[pageSize](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize) is more than 20 and for columns if more than 20 are outside the current viewport.

Reviewer 3: ```suggestion
The DataGrid renders all rows and columns once the data is loaded. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. These properties specify that UI elements are only rendered when they come into the viewport. Use virtual row rendering when [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is more than the default value. Virtual column rendering can be used when there are more than 20 columns outside the viewport.
```

> * **Render certain elements after others**           

Reviewer 3: ```suggestion
* **Render elements with complex content last**           
```

> The DataGrid renders all rows and columns once it loads data. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. In this case, UI elements are only rendered when they come into the viewport. Use virtual row rendering if [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is higher than the default value. Virtual column rendering should be used in grids with more than 20 columns outside the viewport.

Reviewer: ```suggestion
The DataGrid renders all rows and columns once it loads data. To improve performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. In this case, UI elements are only rendered when they come into the viewport. Use virtual row rendering if [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is higher than the default value. Virtual column rendering should be used in grids with more than 20 columns outside the viewport.
```

> The DataGrid renders all rows and columns once the data is loaded. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. These properties specify that UI elements are only rendered when they come into the viewport. Use virtual row rendering when [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is more than the default value. Virtual column rendering should be used when there are more than 20 columns outside the viewport.

Reviewer 2: A few minor edits. One thing I can point out is to always look out for "there is / there are" - these are best avoided. They add wordiness and, like passive voice, help hide clear subject/verb pair (who exactly does what).

```suggestion
The DataGrid renders all rows and columns once it loads data. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. In this case, UI elements are only rendered when they come into the viewport. Use virtual row rendering if [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is higher than the default value. Virtual column rendering should be used in grids with more than 20 columns outside the viewport.
```

> The DataGrid renders all rows and columns once the data is loaded. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. These properties specify that UI elements are only rendered when they come into the viewport. Use virtual row rendering when [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is more than the default value. Virtual column rendering can be used when there are more than 20 columns outside the viewport.

Reviewer 4: Add information about side effects of enabling this options. These options improve render performance after first render and after data operations but can decrease performance during scrolling because content is rendered during scrolling. The renderAsync global and column options can improve performance during scrolling.

> To implement this, enable the [renderAsync](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync) property. It affects filter row, command columns, and columns with [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) set to **true**. You should also enable [this property](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync) for columns with a complex [template](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate). Ensure that they have a [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) specified.

Reviewer 3: ```suggestion
Cells with complex content (predefined templates, command columns, editors) can be rendered after cells with simple content. To do that, enable the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use a content-heavy [cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').
```

> Cells with complex content (predefined templates, command columns, editors) can be rendered after cells with simple content. To do that, enable the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use a content-heavy [cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').

Reviewer 4: What is predefined templates? I do not know. I think we should give more specific information: "filter row cells, command columns, checkboxes in boolean columns". 
Also, in React users use cellRender and cellComponent instead of cellTemplate. Can we say about this react specific?

> Cells with complex content (filter row, command columns, editors and columns with checkboxes) can be rendered after cells with simple content. To do that, enable the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use a content-heavy [cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').

Reviewer 2: The first sentence can be turned into active form. This will bring the subject and verb together at the beginning of the sentence for clearer flow. 

There's not explicit indication of the effect that this rendering strategy has on end-users. You might want to add a sentence in the beginning of after the first one that explains why that matters. I could be clear to many, but still it makes sense to state your point exactly, so that no doubt remains. 

'They use a cellTemplate' - you go from plural to singular (probably to match the property name). If it plural "columns" then it should be "templates". So I changed that part a bit, including the link text. 

```suggestion
The grid control can render cells with simple content first, and only then proceed to cells with complex content (filter row, command columns, editors and columns with checkboxes). If the control does this, end-users start to see content sooner and thus perceive the app UI as more responsive. To load simple content first, use the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use content-heavy [cell templates](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').
```

> The grid control can render cells with simple content first, and only then proceed to cells with complex content (filter row, command columns, editors and columns with checkboxes). If the control does this, end-users start to see content sooner and thus perceive the app UI as more responsive. To load simple content first, use the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use content-heavy [cell templates](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').

Reviewer: ```suggestion
The grid control can render cells with simple content first, and only then proceed to cells with complex content (filter row, command columns, editors and columns with checkboxes). In this case, the control shows content sooner and the app UI can be perceived as more responsive. To load simple content first, use the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. You can also enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use content-heavy [cell templates](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').
```

* **Prevent columns from adjusting their widths to the content**                    

> Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. To optimize rendering even further, set fixed column widths. Specify the fixed width in the column's [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property or use the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) property to do this for all grid columns. Note that command columns ignore columnWidth, and you have to use the individual setting.

Reviewer 3: ```suggestion
Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. To optimize rendering even further, set fixed column widths. Specify the fixed width in the column's [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property or use the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) property to do this for all grid columns. Note that command columns ignore **columnWidth**, and you have to use the individual setting.
```

> Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. To optimize rendering even further, set fixed column widths. Specify the fixed width in the column's [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property or use the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) property to do this for all grid columns. Note that command columns ignore columnWidth, and you have to use the individual setting.

Reviewer 4: Setting fixed column widths for all column or setting columnWidth options does not improve performance if there are command columns. User should assign fixed widths for all command columns the columns option. Without full information these information will be wrong for many cases. So, we should extend information or exclude information about fixed widths from this paragraph and say about disabling columnAutoWidth only.

> Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. You can also set a custom width for all or individual columns in the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) and **columns[]**.[width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) properties respectively. If **columnWidth** is set, specify widths for all command columns - this property does not affect them.

Reviewer 3: ```suggestion
Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled.
```

* **Do not use checkboxes in Boolean columns**          

> Columns with *"boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) use checkboxes to display column values. To disable this feature, assign **false** to the columns' [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property. The checkboxes will be replaced with the **true** and **false** text values.

Reviewer 4: Add information about trueText and falseText column options.

> Columns with *"boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) use checkboxes to display column values. To use the **true** and **false** text values instead, assign **false** to the columns' [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property.

Reviewer 3: ```suggestion
Columns with *"boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) use checkboxes to display column values. To disable this feature, assign **false** to the columns' [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property. The checkboxes will be replaced with the **true** and **false** text values.
```

> Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. To optimize rendering even further, set fixed width for all columns. Specify it in the column's [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property or use the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) property to do this for all columns. Note that command columns ignore **columnWidth**, and you have to use the individual setting.

Reviewer: ```suggestion
Do not enable [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property. To further optimize rendering, use the column's [width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property or [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) property to specify a fixed width for all columns. Note that command columns ignore **columnWidth**, and you should use individual settings.
```

* **Implement [onCellPrepared](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared) instead of [cellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) for conditional formatting**        

> You should do this in Angular, React, or Vue applications only. jQuery applications are compatible with both properties.

Reviewer 3: ```suggestion
Due to the design of DevExtreme UI components, **onCellPrepared** works faster when you need to stylize cells based on a certain condition.
```

> Columns with *"boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) use checkboxes to display column values. To disable this feature, assign **false** to the columns' [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property. The checkboxes will be replaced with the **true** and **false** text values. You can change them in the [trueText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#trueText) and [falseText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#falseText) properties.

Reviewer 2: Boolean should always be capitalized - it comes from somebody's last name. 

Again - there's no sentence that explicitly states that check boxes require more rendering resources than text. You give no clear reason why developers should disable them.

You also seem to be avoiding the phrase "by default", which is per guidelines... but you simply remove it from the sentence, and thus change the meaning of the sentence. You can still use the word "default", only avoid comma-separated clauses in the beginning of your sentences. 

```suggestion
Check boxes are the default display style for columns with *"Boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType). You can increase rendering speed if you instruct the grid to render text instead. To do this, assign **false** to the column's [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property. Use the [trueText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#trueText) and [falseText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#falseText) properties to specify display strings. The default strings are **true** and **false**. 
```

> The default editor for columns of *"Boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) is checkbox. You can increase rendering speed if you instruct the grid to render text instead. To do this, assign **false** to the column's [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways). Use the [trueText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#trueText) and [falseText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#falseText) properties to specify display strings. The default values are **true** and **false**.  

Reviewer: ```suggestion
The checkbox is the default editor for *"Boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) columns. You can increase rendering speed if you instruct the grid to render text instead. To do this, assign **false** to the column's [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways). Use the [trueText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#trueText) and [falseText](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#falseText) properties to specify display strings. The default values are **true** and **false** respectively.  
```

> Due to the design of DevExtreme UI components, **onCellPrepared** works faster when you need to stylize cells based on a certain condition.

Reviewer: `works faster` compared to what? Why does it work faster?

We recommend that you apply the following settings to improve performance:

* **Make certain elements render after others**           
To implement this, enable the [renderAsync](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync) property. It affects filter row, command columns, and columns with [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) set to **true**.

* **Make columns with a complex [template](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) render after others**               
For this, enable the **columns[]**.[renderAsync](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync) property.

* **Prevent columns from adjusting their widths to the content**                    

> For this, disable the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property if enabled. Additionally, you can set a custom width for all or individual columns in the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) and **columns[]**.[width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) properties respectively.

Reviewer 3: "For this," - this has become a pattern encountered in many of your documents. Try to avoid such things.
"Additionally ... respectively" - we start and end a sentence with adverbs, that's not good.

```suggestion
Keep the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property disabled. You can also set a custom width for all or individual columns in the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) and **columns[]**.[width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) properties respectively.
```

> If you work with a large data set, we recommend that you apply these settings to improve rendering performance:

Reviewer 3: "... apply these settings ..." - list items here are not settings.
"... to improve rendering performance ..." - the word "improve" means that something was in bad condition and then it became better. Generally, prefer "enhance". In this example, "optimize" fits even better.

```suggestion
If you work with a large data set, apply the following recommendations to optimize rendering performance:
```

* **Render rows and columns of the current viewport only**             

> The default behavior is to render all loaded rows and columns at once. To render those of the current viewport only, set the [rowRenderingMode](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/scrolling/#rowRenderingMode) and [columnRenderingMode](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/scrolling/#columnRenderingMode) properties to *"virtual"*. You should do this for rows if [paging was disabled](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/paging/#enabled) or the **paging**.[pageSize](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/paging/#pageSize) is more than 20 and for columns if more than 20 are outside the current viewport.

Reviewer 3: Pay attention to the URLs.

* **Render certain elements after others**           
To implement this, enable the [renderAsync](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync) property. It affects filter row, command columns, and columns with [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) set to **true**.

* **Render columns with a complex [template](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) after others**               

> UI component's **renderAsync** property does not affect such columns - enable the **columns[]**.[renderAsync](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync) property for them.

Reviewer 3: 1. Looks like these two items cover the same ground and should be collected together.
2. What are the drawbacks of enabling renderAsync? Based on the description alone, async rendering looks like a cool feature that should be enabled by default. However, that isn't so, and the docs doesn't explain why.

Author: renderAsync was introduced after the synchronous render => enabling this option would result in a BC.

* **Prevent columns from adjusting their widths to the content**                    
For this, disable the [columnAutoWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnAutoWidth) property if enabled. Additionally, you can set a custom width for all or individual columns in the [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) and **columns[]**.[width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) properties respectively.

* **Specify widths for all [command columns](/Documentation/Guide/UI_Components/DataGrid/Columns/Column_Types/Command_Columns/)**         

> If the **columns[]**.[width](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#width) property is not set yet, specify it because [columnWidth](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#columnWidth) does not affect command columns.

Reviewer 3: The items about width should be collected together.

> * **Substitute checkboxes for *true* or *false* text values in columns of [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) *"boolean"***          

Reviewer 3: ```suggestion
* **Do not use checkboxes in Boolean columns**          
```

> For this, disable the **columns[]**.[showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property.

Reviewer 3: ```suggestion
Columns with *"boolean"* [dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) use checkboxes to display column values. To disable this feature, assign **false** to the columns' [showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) property. The checkboxes will be replaced with the **true** and **false** text values.
```

> * **Implement [onCellPrepared](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onCellPrepared) instead of [cellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) for conditional formatting**        

Reviewer 3: Explain the reasons.
